linux/devdoc/completion_port_linux_requirements.md

```diff
@@ -0,0 +1,180 @@
+MOCKABLE_FUNCTION(, int, completion_port_post_queue, COMPLETION_PORT_HANDLE, completion_port, const COMPLETION_PORT_INGRESS_DATA*, ingress_data);
+```
+
+`completion_port_post_queue` adds an ingress item to the post queue for epoll signaling
+
+If `completion_port` is `NULL`, `completion_port_post_queue` shall return a non-NULL value.
+
+If `ingress_data` is `NULL`, `completion_port_post_queue` shall return a non-NULL value.
+
+`completion_port_post_queue` shall allocate a thread data object
+
+`completion_port_post_queue` shall allocate a thread data object
+
+`completion_port_post_queue` shall add the socket in the epoll system by calling `epoll_ctl` with `EPOLL_CTL_MOD` along with the `COMPLETION_PORT_INGRESS_DATA` `epoll_op` variable.
+
+If the `epoll_ctl` call fails with `ENOENT`, `completion_port_post_queue` shall call `epoll_ctl` again with `EPOLL_CTL_ADD`.
+
+On success, `completion_port_post_queue` shall return 0.
+
+If any error occurs, `completion_port_post_queue` shall fail and return a non-zero value.
+
+### completion_port_dequeue
+
+```C
+MOCKABLE_FUNCTION(, void, completion_port_dequeue, COMPLETION_PORT_HANDLE, completion_port, SOCKET_HANDLE, socket);
+```
+
+If `completion_port` is `NULL`, `completion_port_dequeue` shall return.
+
+If `socket` is `INVALID_SOCKET`, `completion_port_dequeue` shall return.
+
+`completion_port_dequeue` shall remove the underlying socket from the epoll by calling  `epoll_ctl` with `EPOLL_CTL_DEL`.
+
+### epoll_worker_func
+
+```c
+static int epoll_worker_func(void* parameter)
+```
+
+If `parameter` is `NULL`, `epoll_worker_func` shall do nothing.
+
+`epoll_worker_func` shall call `epoll_wait` to wait for an epoll event to become signaled.
+
+If the events value contains `EPOLLRDHUP` (hang up), `epoll_worker_func` shall the following:
+
+- `epoll_worker_func` shall receive the `EPOLL_THREAD_DATA` value from the ptr variable from the `epoll_event` data ptr.
+
+- Then call the `event_callback` callback with the `event_callback_ctx` and `COMPLETION_PORT_RESULT_ABANDONED`.
+
+If the events value contains `EPOLLIN` (recv), `epoll_worker_func` shall the following:
+
+- `epoll_worker_func` shall receive the `EPOLL_THREAD_DATA` value from the ptr variable from the `epoll_event` data ptr.
+
+- Then `epoll_worker_func` shall call `recv` and do the following:
+
+- If the recv size < 0, then:
+
+  - If `errno` is `EAGAIN` or `EWOULDBLOCK`, then unlikely errors will continue
+
+  - If `errno` is `ECONNRESET`, then `epoll_worker_func` shall call the `event_callback` callback with the `event_callback_context` and `COMPLETION_PORT_RESULT_ABANDONED`
+
+- If the recv size equal 0, then `epoll_worker_func` shall call `event_callback` callback with the `event_callback_context` and `COMPLETION_PORT_RESULT_OK`
+
+- If the recv size > 0, if we have another buffer to fill then we will attempt another read, otherwise we shall call `event_callback` callback with the `event_callback_context` and `COMPLETION_PORT_RESULT_OK`
```

Comment:
>if we have another buffer to fill [](http://example.com/codeflow?start=24&length=33)

so what happens if we "recv" and we DO NOT have another buffer? Data is dropped?

Reply:
rewording